[SR-510] diagnose enum raw value in closure parsing

dduan opened the pull request:
This reverts the previous fix "complain for invalid enum raw value" and fixes
SR-510 with a different approach:

~~enum raw value is parsed as a normal expression. When a closure is encountered
as the expression, a proper context will be absent. If we pass a diagnosis
from raw value to `parseExprClosure`, it will get properly issued.~~

enum raw value is parsed as a normal expression using `parseExpr()`. However,
for a closure, the parser expects a local context that doesn't exist for raw
values.

We create a temporary context to ensure the closure gets parsed as normal.
As a consequence, `parseExpr()` returns normally for closure and correct
diagnosis for raw value gets issued.

@jrose-apple

jrose-apple commented:
Hm. First request: please split the revert out as a separate commit. :-) (Same pull request is fine, so we don't regress on master, but it is logically separate.)

Beyond that, the error message doesn't really apply here. There *is* an expression there; it's just not one that belongs there.

dduan committed:
revert 'complain for invalid enum raw value'

and all changes introduced in the same pull request

dduan committed:
[SR-510] Make closure as enum raw value parses as normal

enum raw value is parsed as a normal expression using `parseExpr()`. However,
for a closure, the parser expects a local context that doesn't exist for raw
values.

We create a temporary context to ensure the closure gets parsed as normal.
As a consequence, `parseExpr()` returns normally for closure and correct
diagnosis for raw value gets issued.

dduan commented:
@jrose-apple done and done.

dduan commented:
The error message is now "raw value for enum case must be a literal"

jrose-apple commented:
This approach seems reasonable, although the hard-coding under the `case '{'` makes me a little sad.

Do you really think it's better to throw away the closure, rather than to create a dummy local context, parse it, and then complain?

jrose-apple commented:
I guess we can add a very heavy assertion that *some* error was emitted if we end up in the "enum case was null" path. We can't guarantee that it happened while trying to parse the case, but if there haven't been any *other* errors, we won't let this one slip by.

dduan commented:
>  the hard-coding under the case '{' makes me a little sad.

I don't feel good about that neither, it's a hack at best.

> Do you really think it's better to throw away the closure, rather than to create a dummy local context, parse it, and then complain?

We want to complain about the raw value being a closure, not what's wrong in the closure itself, no?

> I guess we can add a very coarse assertion that some error was emitted if we end up in the "enum case was null" path.

That's basically what's in master now. For `case = {}`, we get both "expect expression after =" and "non-literal" diagnosis (not these exact phrases, obviously).

> …but if there haven't been any other errors…

I couldn't figure out how to do this. Tips?

A couple of scenarios with this change applied:

```
$ cat /tmp/t.swift
enum X: String {
    case A
    case B = {
$ ./swift /tmp/t.swift
/tmp/t.swift:3:14: error: raw value for enum case must be a literal
    case B = {
             ^
/tmp/t.swift:4:1: error: expected declaration
```

```
$ cat /tmp/t.swift
enum X: String {
    case A
    case B =
$ ./swift /tmp/t.swift
/tmp/t.swift:3:13: error: expected expression after '=' in 'case'
    case B =
            ^
/tmp/t.swift:4:1: error: expected declaration

^
```

dduan commented:
@jrose-apple the cosmetic changes has been made.

dduan commented:
@jrose-apple `is` only exists for `swift::Diagnostic`, so I can't use it on `swift::Diag` 🙁

dduan commented:
@jrose-apple running test on coarse assertion for null raw value now.

dduan commented:
I tried adding an assertion to raw value's null branch. It caused a few assertion failures in some tests regarding name binding.

dduan commented:
Ok, assertion added.

dduan commented:
Update for @jrose-apple:

I implemented your suggestion. Adding a local context made the patch much simpler. The closure gets parsed properly and errors related to both the closure itself and raw value gets issued properly.

An assertion for the "enum case was null" path is also added.

All tests are passing.

nuclearace commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -4413,7 +4412,7 @@ ParserStatus Parser::parseDeclEnumCase(ParseDeclOptions Flags,
       }
       if (RawValueExpr.isNull()) {
-        diagnose(NextLoc, diag::nonliteral_enum_case_raw_value);
+        assert("No error emmited for invalid enum raw value expression");
```
~~emmited~~ emitted?

dduan replied:
Thank Erik. Fixed.

jrose-apple commented:
Um. We can work out the assertion failures separately (it still seems valuable). I actually meant creating a local context ahead of time, so that we wouldn't have to hardcode a particular ID in the main expr-parsing loop. I guess this is cheaper, though, since it only comes up on failure. Would it at least make sense to move it into the closure-parsing logic, though?

Also, how does that local context ever get cleaned up? Seems like it's just new'ed into space.

dduan commented:
@jrose-apple Ok, I removed the hard-coded diagnosis from `parseExpr`, tests are parsing.

jrose-apple commented:
That still leaks the local context, though, doesn't it?

dduan commented:
@jrose-apple Oops, fixed.

dduan commented:
📣

dduan commented:
@jrose-apple anything else I can do to make this one better?

jrose-apple commented:
Sorry I hadn't gotten back to this! I think this is pretty good. It's not a completely beautiful solution but it does mean we're parsing the same way as everything else. I have nitpicks on your comments but I'll merge it after that.

dduan commented:
Done. Thank you for all the help 😬.
